server/controllers/auth.js

```diff
@@ -52,12 +55,48 @@ function login(req, res, next) {
     WHERE user.username = ? AND user.password = PASSWORD(?) AND project_permission.project_id = ?;
   `;
 
-  db.exec(sql, [username, password, projectId])
-    .then(function (rows) {
+  let sqlUser = `
+    SELECT user.id FROM user
+    WHERE user.username = ? AND user.password = PASSWORD(?)`;
+
+  let sqlProject = `
+    SELECT project_permission.id FROM project_permission 
+    JOIN user ON user.id = project_permission.user_id
+    WHERE user.username = ? AND user.password = PASSWORD(?)`;
+
+  let sqlPermission = `
+    SELECT permission.id FROM permission
+    JOIN user ON user.id = permission.user_id
+    WHERE user.username = ? AND user.password = PASSWORD(?)`;  
+
+  transaction = db.transaction();
+
+  transaction
```

Comment:
MySQL transactions are very useful for when we need to roll back changes, however they are expensive to perform as they lock MySQL tables during execution. 

As all of these operations only involve reads I recommend that a transaction is not used. 

Instead this section could read like this: 

```js
db.exec(sqlTestUser) 
  .then((result) => { 
    // comment describing was result was
   if (result.length) { throw new BadRequest() }
   return db.exec(sqlTestProject);
  })
  .then // ...
```